### ISSUE TITLE
Support for cartesian products?

### ISSUE THREAD

**Issue by rhalbersma** [ericniebler/range-v3#173]
The `meta` library contains a `cartesian_product` for variadic parameter packs. It would be nice if `range-v3` would also support a view for Cartesian products so that a `tuple<Range<T>...>` could be transformed into a `Range<tuple<T...>>`.

**Comment by ericniebler** [comment]
For two ranges, you can use `view::for_each`:

```c++
std::vector<int> v{1,2,3};
std::vector<char> c{'a','b','c'};

auto cross = view::for_each(v, [&c](int i) {
  return view::for_each(c, [i](char c) {
    return yield(Pair<int,char>{i,c});
  });
});
std::cout << cross << std::endl;
```

**Comment by rhalbersma** [comment]
OK, but the variadic case seems a bit unwiedly without using fold expressions, but that's c++1z. Also, I am thinking about how to do the case of a finite number of infinite ranges, doing Cantor's zig-zag of enumerating the rational numbers. I.e. { 0, 1, ... } x { 0, 1, ... } -> {{0,0}, {1, 0}, {0,1}, {2,0}, {1,1}, {0,2}, ...} etc.

**Comment by Predelnik** [comment]
I support it, this range seems pretty natural since it replaces several nested loops and allows algorithm application on items iterated that way.

**Comment by CaseyCarter** [comment]
I have an implementation (https://github.com/CaseyCarter/range-v3/blob/cartesian_product/include/range/v3/view/cartesian_product.hpp) sitting on a branch waiting for more test coverage. If someone wants to contribute some tests, it would be much appreciated.

**Comment by CaseyCarter** [comment]
Should be closed by fd2de9707ff35a346d20e16fb489f80a68bd717c; I forgot to mention this issue in the commit log.

**Comment by CaseyCarter** [comment]
> Also, I am thinking about how to do the case of a finite number of infinite ranges, doing Cantor's zig-zag of enumerating the rational numbers

I don't think this would be hard to implement, but someone needs to convince me that there are motivating use cases for this. Well, motivating enough for me to invest the time to implement it, anyway.
